File: mvp-workflow-app/backend/utils/workflow_executor.py
from ..models.workflow import Workflow
import asyncio

class WorkflowExecutor:
    @staticmethod
    async def execute_workflow(workflow: Workflow):
        for step in workflow.flow:
            await WorkflowExecutor.execute_step(step['node'])

    @staticmethod
    async def execute_step(step: dict):
        node = step['node']
        print(f"ステップを実行中: {node['nodeName']}")
        
        # ノードタイプに基づいて処理を実行
        if node['nodeType'] == 'start':
            await WorkflowExecutor.execute_start_node(node)
        elif node['nodeType'] == 'llm':
            await WorkflowExecutor.execute_llm_node(node)
        elif node['nodeType'] == 'end':
            await WorkflowExecutor.execute_end_node(node)
        # 他のノードタイプの処理を追加

        print(f"ステップ完了: {node['nodeName']}")

    @staticmethod
    async def execute_start_node(node: dict):
        # スタートノードの処理
        await asyncio.sleep(1)

    @staticmethod
    async def execute_llm_node(node: dict):
        # LLMノードの処理
        input_file = node['nodeParameter'].get('inputFileParam')
        max_summary_length = int(node['nodeParameter'].get('maxSummaryLength', 100))

        # ファイルの読み込み処理（実際の実装が必要）
        file_content = await WorkflowExecutor.read_file(input_file)
        
        # LLMを使用したファイル要約処理（実際の実装が必要）
        summary = await WorkflowExecutor.summarize_text(file_content, max_summary_length)
        
        # 結果を保存
        node['output'] = {'summary': summary}

    @staticmethod
    async def execute_end_node(node: dict):
        # エンドノードの処理
        await asyncio.sleep(1)

    @staticmethod
    async def read_file(file_path):
        # ファイル読み込み処理を実装（仮の実装）
        return f"ファイル {file_path} の内容"

    @staticmethod
    async def summarize_text(text, max_length):
        # LLMを使用したテキスト要約処理を実装（仮の実装）
        return f"{text[:max_length]}... (要約: {len(text)}文字)"

File: mvp-workflow-app/backend/src/utils/workflow_executor.ts
import { API } from "../models/api";
import { AIService } from "../services/ai_service";

export class WorkflowExecutor {
  static async executeWorkflow(api: API, inputText?: string): Promise<any> {
    let result: any = { text: inputText || "Sample input text" };
    if (Array.isArray(api.flow)) {
      for (const step of api.flow) {
        if (step && step.node) {
          result = await this.executeStep(step.node, result);
        } else {
          console.error("Invalid step in api.flow:", step);
        }
      }
    } else {
      console.error("api.flow is not an array or is undefined:", api.flow);
    }
    return result;
  }

  private static async executeStep(node: any, input: any): Promise<any> {
    console.log(`Executing step: ${node.nodeName}`);
    switch (node.nodeType) {
      case "start":
        return { text: input.text || "Sample input text" };
      case "llm":
        return this.executeLLMNode(node, input);
      case "end":
        return input;
      default:
        throw new Error(`Unknown node type: ${node.nodeType}`);
    }
  }

  private static async executeLLMNode(node: any, input: any): Promise<any> {
    switch (node.nodeParameter.aiFunction) {
      case "summarize":
        const summary = await AIService.summarizeText(
          input.text,
          node.nodeParameter.maxLength || 100
        );
        return summary;
      // 他のAI機能も同様に実装できます
      default:
        throw new Error(
          `Unknown AI function: ${node.nodeParameter.aiFunction}`
        );
    }
  }
}
